README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# dammmdatagen <a href="https://doktormike.github.io/dammmdatagen/"><img src="man/figures/logo.png" align="right" height="139" /></a>

The goal of dammmdatagen is to make it easy for marketing mix modeling
professionals to get access to realistic data sets where the ground
truth is known. This fascilitates our development and provides in the
end more value for all stakeholders of MMM.

## Build status etc

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/DoktorMike/dammmdatagen/workflows/R-CMD-check/badge.svg)](https://github.com/DoktorMike/dammmdatagen/actions)
[![Codecov test
coverage](https://codecov.io/gh/DoktorMike/dammmdatagen/branch/master/graph/badge.svg)](https://app.codecov.io/gh/DoktorMike/dammmdatagen?branch=master)
<!-- badges: end -->

## Installation

You can install dammmdatagen from github with:

``` r
# install.packages("devtools")
devtools::install_github("DoktorMike/dammmdatagen")
```

## Quick start

This is a basic example which shows you how to generate a small 1 year
data set.

``` r
# load useful libraries
library(dammmdatagen)

# generate a basic data set
mydf <- generateCovariatesData()
#> Joining, by = "date"
#> Joining, by = "date"
#> Joining, by = "date"
head(mydf)
#> # A tibble: 6 × 37
#>   date       sunshine precipitation temperature competitor_a competitor_b
#>   <date>        <dbl>         <dbl>       <dbl>        <dbl>        <dbl>
#> 1 2020-12-18 -0.729          -1.66        0.118        60000        20000
#> 2 2020-12-19 -0.197          -1.87        0.322       100000        40000
#> 3 2020-12-20  1.04           -1.37       -1.27         40000            0
#> 4 2020-12-21  0.157           0.457       0.231        50000        10000
#> 5 2020-12-22  0.00165         1.70        1.55         40000        90000
#> 6 2020-12-23 -0.311           0.995       0.719        10000        60000
#> # … with 31 more variables: competitor_c <dbl>, cpi <dbl>, cci <dbl>,
#> #   gdp <dbl>, dist_product_a <dbl>, dist_product_b <dbl>,
#> #   dist_product_c <dbl>, price_product_a <dbl>, price_product_b <dbl>,
#> #   price_product_c <dbl>, net_display <dbl>, net_facebook <dbl>,
#> #   net_search_branded <dbl>, net_tv <dbl>, net_radio <dbl>, net_ooh <dbl>,
#> #   net_print <dbl>, imp_display <dbl>, imp_facebook <dbl>,
#> #   imp_search_branded <dbl>, imp_tv <dbl>, imp_radio <dbl>, imp_ooh <dbl>, …
```

Say that you would like to also generate a response variable to fit a
model to. Then you could use the highlevel API function below.

``` r
library(dammmdatagen)
library(ggplot2)
library(dplyr)
library(tidyr)
library(scales)
ret <- generateRetailData()
#> Joining, by = "date"
#> Joining, by = "date"
#> Joining, by = "date"
#> New names:
#> * product_a -> product_a...7
#> * product_b -> product_b...8
#> * product_c -> product_c...9
#> * product_a -> product_a...10
#> * product_b -> product_b...11
#> * ...
dates <- ret[["covariates"]][["Macro"]][["date"]]
qplot(dates, ret[["response"]]) + geom_line() + ylim(0, NA)
```

<img src="man/figures/README-fullexample-1.png" width="100%" />

``` r
# entrytocolname <- function(x) a <- ret[["effects"]][[x]] %>% setNames(c(tolower(paste0(x, "_", names(.)))))
entrytocolname <- function(x) tibble::tibble(rowSums(ret[["effects"]][[x]])) %>% setNames(x)
Reduce(dplyr::bind_cols, lapply(names(ret[["effects"]]), entrytocolname)) %>%
        dplyr::mutate(date = dates) %>%
        tidyr::pivot_longer(-date, names_to = "variable", values_to = "value") %>%
        ggplot2::ggplot(ggplot2::aes(x = date, y = value, fill = variable)) +
    ggplot2::geom_bar(stat = "identity") + ggplot2::theme_minimal() +
    ggplot2::ylab("Units sold") +
    ggplot2::xlab("")
```

<img src="man/figures/README-fullexample-2.png" width="100%" />

We can do a lot more of course! In this small snippet we’ll generate 1
month worth of competitor media spendings data and plot that out.

``` r
library(dammmdatagen)
library(ggplot2)
library(dplyr)
library(tidyr)
library(scales)

generateCompetitorData(fromDate = Sys.Date() - 30, toDate = Sys.Date()) %>%
        gather("competitor", "spend", -"date") %>%
        ggplot(aes(y = spend, x = date, fill = competitor)) +
        geom_bar(stat = "identity", position = position_stack()) +
        theme_minimal() +
        scale_y_continuous(labels = dollar_format(prefix = "kr. "))
```

<img src="man/figures/README-competitorspendplot-1.png" width="100%" />

Just as we can generate competitor spending data we can also generate
macroeconomical data. These types of indicators are typically slow
moving over time with minor temporal differences.

``` r
generateMacroData(fromDate = Sys.Date() - 30, toDate = Sys.Date()) %>%
        gather("indicator", "value", -"date") %>%
        ggplot(aes(y = value, x = date, color = indicator)) +
        geom_line(size = 1.5) +
        theme_minimal()
```

<img src="man/figures/README-macroecondataplot-1.png" width="100%" />

## Event type data

Event data are modeled as a poisson distribution with a low incidence.

``` r
generateEventData(Sys.Date() - 265, Sys.Date()) %>%
        gather(type, value, -date) %>%
        ggplot(aes(y = value, x = date, fill = type)) +
        geom_bar(stat = "identity") +
        theme_minimal()
```

<img src="man/figures/README-eventdata1-1.png" width="100%" />

The incidence can of course be controlled. This is done via the freq
parameter.

``` r
generateEventData(Sys.Date() - 265, Sys.Date(), freq = 0.1) %>%
        gather(type, value, -date) %>%
        ggplot(aes(y = value, x = date, fill = type)) +
        geom_bar(stat = "identity") +
        theme_minimal()
```

<img src="man/figures/README-eventdata2-1.png" width="100%" />

## Media generation

Generating media is in general a bit more complicated as we need more
information since in MMM models that’s what we primarily care about. So
we need three data.frames; the net, the impressions and the cpms. We
also differentiate between offline and online media. This difference is
rather artificial right now but it’s to futureproof the package.

``` r
mydflist <- generateOnlineData(Sys.Date() - 30, Sys.Date())
mydflist[["impression"]] %>%
        gather(type, impression, -date) %>%
        ggplot(aes(y = impression, x = date, fill = type)) +
        geom_bar(stat = "identity") +
        theme_minimal()
```

<img src="man/figures/README-onlineimpdata-1.png" width="100%" />

## Code of Conduct

Please note that the dammmdatagen project is released with a
[Contributor Code of
Conduct](https://doktormike.github.io/dammmdatagen/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.0.0 (2021-12-27)


### Features

* Added a feature for generating a complete response based on a retail scenario with massive base shifts. ([57ea494](https://github.com/DoktorMike/dammmdatagen/commit/57ea4940da55719bab766a5de9b1b23c1082aa80))


### Bug Fixes

* fixed the build process to make sure we can execute examples with the pipe operator. ([2af4dfe](https://github.com/DoktorMike/dammmdatagen/commit/2af4dfe1f89075a605abc6c4088ca6bf43cb6ab6))


tests/testthat.R
library(testthat)
library(dammmdatagen)

test_check("dammmdatagen")


tests/testthat/test_simple.R
# library(stringr)
context("Dimensions check on data")

offonsize <- structure(c(366L, 4L, 366L, 4L, 366L, 4L),
  .Dim = 2:3,
  .Dimnames = list(NULL, c("net", "impression", "cpm"))
)

test_that("we get the correct dimensions", {
  expect_equal(dim(generateMacroData()), c(366, 4))
  expect_equal(dim(generateEventData()), c(366, 3))
  expect_equal(dim(generateWeatherData()), c(366, 4))
  expect_equal(dim(generateCompetitorData()), c(366, 4))
  expect_equal(dim(generatePriceData()), c(366, 4))
  expect_equal(generateOnlineData() %>% sapply(dim), offonsize)
  expect_equal(generateOfflineData() %>% sapply(dim), offonsize)
})


tests/testthat/test_usecase.R
context("Top level API checks")

ret <- generateRetailData()

test_that("we get three list items back", {
  expect_equal(length(ret), 3)
})


R/effectgen.R
#' Generate a marketing mix modeling data set
#'
#' This function generates a marketing mix modeling data set based on the
#' parameters given for a Retail case with Revenue as it's response.
#' It uses many different stochastical processes to accomplish
#' this and the dynamics behind them are not available to the user to manipulate.
#'
#' @importFrom magrittr "%>%"
#'
#' @param fromDate the beginning of the time series
#' @param toDate the end of the time series
#' @param kpi the name of the kpi (response) to simulate
#' @param sector the name of the sector to simulate (currently not used)
#' @param onlineInsertionNames the names of each online media insertion you wish to use
#' @param offlineInsertionNames  the names of each offline media insertion you wish to use
#' @param priceNames the names of the different product prices
#' @param distributionNames the names of the different product distributions
#' @param weatherNames the names of the weather data measurements to use
#' @param competitorNames the names of each competitor considered
#' @param macroNames the names of the macroeconomical factors to simulate
#' @param eventNames the names of events to add
#'
#' @return a list of tibble containing each data mentioned
#' @export
#'
#' @examples
#' library(ggplot2)
#' library(nord)
#' ret <- generateRetailData()
#' dates <- ret[["covariates"]][["Macro"]][["date"]]
#' qplot(dates, ret[["response"]]) + geom_line() + ylim(0, NA)
#' # entrytocolname <- function(x) a <- ret[["effects"]][[x]] %>% setNames(c(tolower(paste0(x, "_", names(.)))))
#' entrytocolname <- function(x) tibble::tibble(rowSums(ret[["effects"]][[x]])) %>% setNames(x)
#' Reduce(dplyr::bind_cols, lapply(names(ret[["effects"]]), entrytocolname)) %>%
#'   dplyr::mutate(date = dates) %>%
#'   tidyr::pivot_longer(-date, names_to = "variable", values_to = "value") %>%
#'   ggplot2::ggplot(ggplot2::aes(x = date, y = value, fill = variable)) +
#'   ggplot2::geom_bar(stat = "identity")
generateRetailData <-
  function(fromDate = Sys.Date() - 3 * 365,
           toDate = Sys.Date(),
           kpi = "units",
           sector = "retail",
           onlineInsertionNames = c("display", "facebook", "search_branded"),
           offlineInsertionNames = c("tv", "radio", "ooh", "print"),
           priceNames = c("price_product_a", "price_product_b", "price_product_c"),
           distributionNames = c("dist_product_a", "dist_product_b", "dist_product_c"),
           weatherNames = c("sunshine", "precipitation", "temperature"),
           competitorNames = c("competitor_a", "competitor_b", "competitor_c"),
           macroNames = c("cpi", "cci", "gdp"),
           eventNames = c("event_a", "event_b")) {
    # GENERATE DATA
    # fromDate <- as.Date("2019-01-01")
    # toDate <- as.Date("2021-12-31")
    datedf <- tibble::tibble(date = seq(fromDate, toDate, by = "1 day"))
    ondf <- generateOnlineData(fromDate, toDate)
    ofdf <- generateOfflineData(fromDate, toDate)
    # Fix them into usable flat tables
    netdf <- dplyr::inner_join(
      ondf$net %>% setNames(c("date", paste0("net_", names(.)[-1]))),
      ofdf$net %>% setNames(c("date", paste0("net_", names(.)[-1])))
    )
    cpmdf <- dplyr::inner_join(
      ondf$cpm %>% setNames(c("date", paste0("cpm_", names(.)[-1]))),
      ofdf$cpm %>% setNames(c("date", paste0("cpm_", names(.)[-1])))
    )
    impdf <- dplyr::inner_join(
      ondf$impression %>% setNames(c("date", paste0("imp_", names(.)[-1]))),
      ofdf$impression %>% setNames(c("date", paste0("imp_", names(.)[-1])))
    )
    # These come as pure tibbles
    prdf <- generatePriceData(fromDate, toDate)
    didf <- generateDistributionData(fromDate, toDate)
    wedf <- generateWeatherData(fromDate, toDate)
    codf <- generateCompetitorData(fromDate, toDate)
    madf <- generateMacroData(fromDate, toDate)
    evdf <- generateEventData(fromDate, toDate)

    # GENERATE EFFECT
    margin <- 400
    roi <- 1.0
    adsf <- function(x, b = 1, a = max(x), l = stats::rbeta(1, 1, 3), p = 1.0) {
      b * tanh(as.vector(stats::filter(x, l, method = "recursive")) / a) * stats::rbinom(1, 1, p)
    }
    linf <- function(x, b = 0, p = 1.0) b * x * stats::rbinom(1, 1, p)
    invf <- function(x, b = 1, a = 1, p = 1.0) b / ((x)^a) * stats::rbinom(1, 1, p)

    ebadf <- tibble::tibble(base = stats::rnorm(nrow(prdf), 500, 10) + 500 * (sin(1:nrow(prdf) / (15 * 2 * pi))))
    enetdf <- plyr::numcolwise(adsf, b = 100, p = 0.7)(netdf)
    eprdf <- plyr::numcolwise(invf, b = 20000, p = 1.0)(prdf)
    edidf <- plyr::numcolwise(linf, b = 0.22, p = 1.0)(didf)
    ewedf <- plyr::numcolwise(linf, b = 100, p = 1.0)(wedf)
    ecodf <- plyr::numcolwise(linf, b = -1e-3, p = 1.0)(codf)
    emadf <- plyr::numcolwise(linf, b = 1, p = 1.0)(madf)
    y <- rowSums(Reduce(dplyr::bind_cols, list(enetdf, eprdf, edidf, ewedf, ecodf, emadf))) + ebadf$base
    # qplot(madf$date, y) + geom_line()
    ret <- list(
      covariates = list(
        Price = prdf, Distribution = didf, Weather = wedf,
        Media = netdf, Competitor = codf, Macro = madf, Events = evdf
      ),
      effects = list(
        Base = ebadf, Price = eprdf, Distribution = edidf, Weather = ewedf,
        Media = enetdf, Competitor = ecodf, Macro = emadf
      ),
      response = y
    )
    ret
  }


R/datagen.R


generateCampaigns <- function(n = 5, l = 30, fromDate = Sys.Date() - 364, toDate = Sys.Date()) {
  dateseq <- seq(fromDate, toDate, by = "1 day")
  numPossibleCampaigns <- as.integer(length(dateseq) / l)
  numPossibleCampaigns

  # library(HMM)
  # Initialise HMM
  # hmm = initHMM(c("Promotion", "Awareness", "NA"),
  #               c("Offline", "Online", "Combined", "NA"),
  #               transProbs=matrix(c(.8,.1,.1,
  #                                   .1,.8,.1,
  #                                   .1,.1,.8), 3),
  #               emissionProbs=matrix(c(.1,.6,.0,
  #                                      .2,.2,.0,
  #                                      .7,.2,.0,
  #                                      .0,.0,1.), 4))
  # print(hmm)
  #
  # # Simulate
  # simHMM(hmm, 30)
  #
  #
  # # Sequence of observations
  # observations = c("L","L","R","R")
  # # Calculate posterior probablities of the states
  # posterior = posterior(hmm,observations)
  # print(posterior)
}

generateFromFunction <- function(f, fromDate = Sys.Date() - 1 * 365,
                                 toDate = Sys.Date(),
                                 mynames = c("something", "something_else")) {
  tmpdf <- tibble::tibble(date = seq(fromDate, toDate, by = "1 day"))
  tmpdf <- data.frame(tmpdf, sapply(mynames, f)) %>% tibble::as_tibble()
  tmpdf
}

#' Generate weather data
#'
#' @param fromDate the beginning of the time series
#' @param toDate the end of the time series
#' @param mynames the names to attach to the generated data
#'
#' @return a tibble with the generated data one column for each element in name
#' @importFrom dplyr "%>%"
#' @importFrom tidyr gather
#' @import ggplot2
#' @export
#'
#' @examples
#' library(ggplot2)
#' library(dplyr)
#' library(tidyr)
#' generateWeatherData(Sys.Date() - 30, Sys.Date()) %>%
#'   gather(type, measure, -date) %>%
#'   ggplot(aes(y = measure, x = date, color = type)) +
#'   geom_line() +
#'   theme_minimal()
generateWeatherData <- function(fromDate = Sys.Date() - 1 * 365,
                                toDate = Sys.Date(),
                                mynames = c("sunshine", "precipitation", "temperature")) {
  arf <- function(x) {
    as.vector(stats::arima.sim(list(order = c(1, 0, 0), ar = 0.7), n = as.integer(toDate - fromDate) + 1))
  }
  generateFromFunction(arf, fromDate = fromDate, toDate = toDate, mynames = mynames)
}

#' Generate price data
#'
#' @param fromDate the beginning of the time series
#' @param toDate the end of the time series
#' @param mynames the names to attach to the generated data
#'
#' @return a tibble with the generated data one column for each element in name
#' @importFrom dplyr "%>%"
#' @importFrom tidyr gather
#' @import ggplot2
#' @export
#'
#' @examples
#' library(ggplot2)
#' library(dplyr)
#' library(tidyr)
#' generatePriceData(Sys.Date() - 30, Sys.Date()) %>%
#'   gather(type, price, -date) %>%
#'   ggplot(aes(y = price, x = date, color = type)) +
#'   geom_line() +
#'   theme_minimal()
generatePriceData <- function(fromDate = Sys.Date() - 1 * 365,
                              toDate = Sys.Date(),
                              mynames = c("product_a", "product_b", "product_c")) {
  arf <- function(x) {
    # Initialise HMM
    hmm <- HMM::initHMM(
      c("PriceWar", "Normal"),
      c("PriceA", "PriceB", "PriceC", "PriceD"),
      transProbs = matrix(c(
        .8, .2,
        .2, .8
      ), 2),
      emissionProbs = matrix(c(
        .3, .6,
        .2, .2,
        .3, .1,
        .2, .1
      ), 4)
    )
    tmptypedf <-
      tibble::tibble(type = HMM::simHMM(hmm, as.integer(toDate - fromDate) +
        1)$observation)
    tmppricedf <-
      tibble::tibble(
        type = c("PriceA", "PriceB", "PriceC", "PriceD"),
        price = c(199, 149, 129, 99)
      )
    tmpdf <- dplyr::left_join(tmptypedf, tmppricedf, by = "type")
    as.vector(tmpdf$price)
  }
  generateFromFunction(arf,
    fromDate = fromDate,
    toDate = toDate,
    mynames = mynames
  )
}

#' Generate distribution data as absolute levels
#'
#' @param fromDate the beginning of the time series
#' @param toDate the end of the time series
#' @param mynames the names to attach to the generated data
#' @param initDist the vector of initial distributions in total numbers
#'
#' @return a tibble with the generated data one column for each element in name
#' @importFrom dplyr "%>%"
#' @importFrom tidyr gather
#' @import ggplot2
#' @export
#'
#' @examples
#' library(ggplot2)
#' library(dplyr)
#' library(tidyr)
#' generateDistributionData(
#'   Sys.Date() - 30, Sys.Date(),
#'   c("product_a", "product_b", "product_c"), c(20, 50, 10)
#' ) %>%
#'   gather(type, distribution, -date) %>%
#'   ggplot(aes(y = distribution, x = date, color = type)) +
#'   geom_line() +
#'   theme_minimal()
generateDistributionData <- function(fromDate = Sys.Date() - 1 * 365,
                                     toDate = Sys.Date(),
                                     mynames = c("product_a", "product_b", "product_c"),
                                     initDist = c(1000, 500, 800)) {
  arf <- function(x) (rpois(as.integer(toDate - fromDate) + 1, 2) / 10 + 1) / 2
  arf <- function(x) {
    # Initialise HMM
    hmm <- HMM::initHMM(
      c("Normal", "Turbulent"),
      c("Increase", "Decrease", "Stay"),
      transProbs = matrix(c(
        .95, .95,
        .05, .05
      ), 2),
      emissionProbs = matrix(c(
        .05, .4,
        .05, .2,
        .90, .4
      ), 3)
    )
    tmptypedf <-
      tibble::tibble(type = HMM::simHMM(hmm, as.integer(toDate - fromDate) +
        1)$observation)
    tmppricedf <-
      tibble::tibble(
        type = c("Increase", "Decrease", "Stay"),
        price = c(1, -1, 0)
      )
    tmpdf <- dplyr::left_join(tmptypedf, tmppricedf, by = "type")
    cumsum(as.vector(tmpdf$price))
  }
  # dautility::qplotez((rpois(10, 2)/10+1)/2)
  a <- generateFromFunction(arf,
    fromDate = fromDate,
    toDate = toDate,
    mynames = mynames
  )
  a[, -1] <- t(t(a[, -1]) + initDist)
  a
}

#' Generate macro economical data
#'
#' @param fromDate the beginning of the time series
#' @param toDate the end of the time series
#' @param mynames the names to attach to the generated data
#'
#' @return a tibble with the generated data one column for each element in name
#' @importFrom dplyr "%>%"
#' @importFrom tidyr gather
#' @import ggplot2
#' @export
#'
#' @examples
#' library(ggplot2)
#' library(dplyr)
#' library(tidyr)
#' generateMacroData(Sys.Date() - 30, Sys.Date()) %>%
#'   gather(type, measure, -date) %>%
#'   ggplot(aes(y = measure, x = date, color = type)) +
#'   geom_line() +
#'   theme_minimal()
#'
#' ts.plot(arima.sim(list(order = c(1, 1, 0), ar = 0.7), n = 365))
generateMacroData <- function(fromDate = Sys.Date() - 1 * 365,
                              toDate = Sys.Date(),
                              mynames = c("cpi", "cci", "gdp")) {
  arf <- function(x) as.vector(stats::arima.sim(list(order = c(1, 1, 0), ar = 0.7), n = as.integer(toDate - fromDate)))
  generateFromFunction(arf, fromDate = fromDate, toDate = toDate, mynames = mynames)
}

#' Generate competitor media spending as list prices data
#'
#' @param fromDate the beginning of the time series
#' @param toDate the end of the time series
#' @param mynames the names to attach to the generated data
#'
#' @return a tibble with the generated data one column for each element in name
#' @importFrom dplyr "%>%"
#' @importFrom tidyr gather
#' @import ggplot2
#' @export
#'
#' @examples
#' library(ggplot2)
#' library(dplyr)
#' library(tidyr)
#' generateCompetitorData(Sys.Date() - 30, Sys.Date()) %>%
#'   gather(competitor, spend, -date) %>%
#'   ggplot(aes(y = spend, x = date, fill = competitor)) +
#'   geom_bar(stat = "identity", position = position_stack())
generateCompetitorData <- function(fromDate = Sys.Date() - 1 * 365,
                                   toDate = Sys.Date(),
                                   mynames = c("competitor_a", "competitor_b", "competitor_c")) {
  arf <- function(x) sample(c(0, seq(10000, 100000, 10000)), as.integer(toDate - fromDate) + 1, replace = TRUE)
  generateFromFunction(arf, fromDate = fromDate, toDate = toDate, mynames = mynames)
}

#' Generate online media data
#'
#' @param fromDate the beginning of the time series
#' @param toDate the end of the time series
#' @param mynames the names to attach to the generated data
#' @param avgcpm the average Cost Per 1000 Impressions
#' @param avgnet the average net investment per insertion and day
#'
#' @return a tibble with the generated data one column for each element in name
#' @importFrom dplyr "%>%"
#' @importFrom tidyr gather
#' @import ggplot2
#' @export
#'
#' @examples
#' library(ggplot2)
#' library(dplyr)
#' library(tidyr)
#' mydflist <- generateOnlineData(Sys.Date() - 30, Sys.Date())
#' mydflist[["impression"]] %>%
#'   gather(type, impression, -date) %>%
#'   ggplot(aes(y = impression, x = date, color = type)) +
#'   geom_line() +
#'   theme_minimal()
generateOnlineData <- function(fromDate = Sys.Date() - 1 * 365,
                               toDate = Sys.Date(),
                               mynames = c("display", "facebook", "search_branded"),
                               avgcpm = 0.5, avgnet = 10000) {
  genCampaignStructure <- function(n) {
    hmm <- HMM::initHMM(
      c("Burst", "Normal", "Off"),
      c("PriceA", "PriceB"),
      transProbs = matrix(c(
        .7, .05, .1,
        .1, .9, .1,
        .2, .05, .8
      ), 3),
      emissionProbs = matrix(c(
        .3, .7, .5,
        .7, .3, .5
      ), 2)
    )
    HMM::simHMM(hmm, n)$states
  }

  arf <- function(x) {
    cpms <- c(seq(0.1 * avgcpm, 2.5 * avgcpm, length.out = 11))
    pricenames <- paste0("Price", LETTERS[1:length(cpms)])
    tmppricedf <- tibble::tibble(cpm = cpms, type = pricenames)
    # Initialise HMM
    hmm <- HMM::initHMM(
      c("High", "Low"),
      pricenames,
      transProbs = matrix(c(
        .7, .3,
        .3, .7
      ), 2),
      emissionProbs = matrix(c(
        .0, .3,
        .0, .1,
        .0, .2,
        .2, .2,
        .2, .2,
        .2, .0,
        .1, .0,
        .1, .0,
        .1, .0,
        .05, .0,
        .05, 0
      ), 11)
    )
    tmptypedf <-
      tibble::tibble(type = HMM::simHMM(hmm, as.integer(toDate - fromDate) +
        1)$observation)
    tmpdf <- dplyr::left_join(tmptypedf, tmppricedf, by = "type")
    as.vector(tmpdf$cpm)
  }
  cpmdf <- generateFromFunction(arf, fromDate = fromDate, toDate = toDate, mynames = mynames)
  campdf <- dplyr::left_join(tibble::tibble(strategy = genCampaignStructure(as.integer(toDate - fromDate) + 1)),
    tibble::tibble(strategy = c("Burst", "Normal", "Off"), net = c(2.5 * avgnet, avgnet, 0)),
    by = "strategy"
  )
  stopifnot(nrow(cpmdf) == nrow(campdf))
  # genspend <- function(x) rnorm(length(x), 1, 1/5)*x
  genimp <- function(x) rnorm(length(x), 1, 1 / 5) * campdf$net / x * 1000
  impdf <- tibble::as_tibble(data.frame(date = cpmdf$date, sapply(dplyr::select(cpmdf, -date), genimp)))
  netdf <- data.frame(
    date = impdf$date,
    dplyr::select(impdf, -date) / 1000 * dplyr::select(cpmdf, -date)
  ) %>%
    tibble::as_tibble()
  list(net = netdf, impression = impdf, cpm = cpmdf)
}

#' Generate offline media data
#'
#' @param fromDate the beginning of the time series
#' @param toDate the end of the time series
#' @param mynames the names to attach to the generated data
#' @param avgcpm the average Cost Per 1000 Impressions
#' @param avgnet the average net investment per insertion and day
#'
#' @return a tibble with the generated data one column for each element in name
#' @importFrom dplyr "%>%"
#' @importFrom tidyr gather
#' @import ggplot2
#' @export
#'
#' @examples
#' library(ggplot2)
#' library(dplyr)
#' library(tidyr)
#' mydflist <- generateOfflineData(Sys.Date() - 30, Sys.Date())
#' mydflist[["impression"]] %>%
#'   gather(type, impression, -date) %>%
#'   ggplot(aes(y = impression, x = date, color = type)) +
#'   geom_line() +
#'   theme_minimal()
generateOfflineData <- function(fromDate = Sys.Date() - 1 * 365,
                                toDate = Sys.Date(),
                                mynames = c("tv", "radio", "ooh"),
                                avgcpm = 0.5, avgnet = 10000) {
  genCampaignStructure <- function(n) {
    hmm <- HMM::initHMM(
      c("Burst", "Normal", "Off"),
      c("PriceA", "PriceB"),
      transProbs = matrix(c(
        .7, .05, .1,
        .1, .9, .1,
        .2, .05, .8
      ), 3),
      emissionProbs = matrix(c(
        .3, .7, .5,
        .7, .3, .5
      ), 2)
    )
    HMM::simHMM(hmm, n)$states
  }

  arf <- function(x) {
    cpms <- c(seq(0.1 * avgcpm, 2.5 * avgcpm, length.out = 11))
    pricenames <- paste0("Price", LETTERS[1:length(cpms)])
    tmppricedf <- tibble::tibble(cpm = cpms, type = pricenames)
    # Initialise HMM
    hmm <- HMM::initHMM(
      c("High", "Low"),
      pricenames,
      transProbs = matrix(c(
        .7, .3,
        .3, .7
      ), 2),
      emissionProbs = matrix(c(
        .0, .3,
        .0, .1,
        .0, .2,
        .2, .2,
        .2, .2,
        .2, .0,
        .1, .0,
        .1, .0,
        .1, .0,
        .05, .0,
        .05, 0
      ), 11)
    )
    tmptypedf <-
      tibble::tibble(type = HMM::simHMM(hmm, as.integer(toDate - fromDate) +
        1)$observation)
    tmpdf <- dplyr::left_join(tmptypedf, tmppricedf, by = "type")
    as.vector(tmpdf$cpm)
  }
  cpmdf <- generateFromFunction(arf, fromDate = fromDate, toDate = toDate, mynames = mynames)
  campdf <- dplyr::left_join(tibble::tibble(strategy = genCampaignStructure(as.integer(toDate - fromDate) + 1)),
    tibble::tibble(strategy = c("Burst", "Normal", "Off"), net = c(2.5 * avgnet, avgnet, 0)),
    by = "strategy"
  )
  stopifnot(nrow(cpmdf) == nrow(campdf))
  # genspend <- function(x) rnorm(length(x), 1, 1/5)*x
  genimp <- function(x) rnorm(length(x), 1, 1 / 5) * campdf$net / x * 1000
  impdf <- tibble::as_tibble(data.frame(date = cpmdf$date, sapply(dplyr::select(cpmdf, -date), genimp)))
  netdf <- data.frame(date = impdf$date, dplyr::select(impdf, -date) / 1000 * dplyr::select(cpmdf, -date)) %>%
    tibble::as_tibble()
  list(net = netdf, impression = impdf, cpm = cpmdf)
}

#' Generate offline media data
#'
#' @param fromDate the beginning of the time series
#' @param toDate the end of the time series
#' @param mynames the names to attach to the generated data
#' @param freq the rate of incidence of events
#'
#' @return a tibble with the generated data one column for each element in name
#' @importFrom dplyr "%>%"
#' @importFrom tidyr gather
#' @import ggplot2
#' @export
#'
#' @examples
#' library(ggplot2)
#' library(dplyr)
#' library(tidyr)
#' generateEventData(Sys.Date() - 30, Sys.Date()) %>%
#'   gather(type, value, -date) %>%
#'   ggplot(aes(y = value, x = date, color = type)) +
#'   geom_line() +
#'   theme_minimal()
generateEventData <- function(fromDate = Sys.Date() - 1 * 365,
                              toDate = Sys.Date(),
                              mynames = c("event_a", "event_b"),
                              freq = 0.01) {
  arf <- function(x) rpois(as.integer(toDate - fromDate) + 1, freq)
  generateFromFunction(arf, fromDate = fromDate, toDate = toDate, mynames = mynames)
}

#' Generate a marketing mix modeling covariates data set
#'
#' This function generates a marketing mix modeling covariates data set based on
#' the parameters given. It uses many different stochastical processes to
#' accomplish this and the dynamics behind them are not available to the user to
#' manipulate. This is supposed to generate all tha variables you use in a
#' marketing mix model except for the response variable i.e. the dependent
#' variable.
#'
#'
#' @param fromDate the beginning of the time series
#' @param toDate the end of the time series
#' @param sector the name of the sector to simulate (currently not used)
#' @param onlineInsertionNames the names of each online media insertion you wish to use
#' @param offlineInsertionNames  the names of each offline media insertion you wish to use
#' @param priceNames the names of the different product prices
#' @param distributionNames the names of the different product distributions
#' @param weatherNames the names of the weather data measurements to use
#' @param competitorNames the names of each competitor considered
#' @param macroNames the names of the macroeconomical factors to simulate
#' @param eventNames the names of events to add
#'
#' @return a list of tibble containing each data mentioned
#' @export
#'
#' @examples
#' a <- 1
generateCovariatesData <-
  function(fromDate = Sys.Date() - 1 * 365,
           toDate = Sys.Date(),
           sector = "retail",
           onlineInsertionNames = c("display", "facebook", "search_branded"),
           offlineInsertionNames = c("tv", "radio", "ooh", "print"),
           priceNames = c("price_product_a", "price_product_b", "price_product_c"),
           distributionNames = c("dist_product_a", "dist_product_b", "dist_product_c"),
           weatherNames = c("sunshine", "precipitation", "temperature"),
           competitorNames = c("competitor_a", "competitor_b", "competitor_c"),
           macroNames = c("cpi", "cci", "gdp"),
           eventNames = c("event_a", "event_b")) {
    mydf <- tibble::tibble(date = seq(fromDate, toDate, by = "1 day"))

    # These come as list of three tibbles.
    ondf <- generateOnlineData(fromDate, toDate, onlineInsertionNames)
    ofdf <- generateOfflineData(fromDate, toDate, offlineInsertionNames)
    # Fix them into usable flat tables
    netdf <- dplyr::inner_join(
      ondf$net %>% setNames(c("date", paste0("net_", names(.)[-1]))),
      ofdf$net %>% setNames(c("date", paste0("net_", names(.)[-1])))
    )
    cpmdf <- dplyr::inner_join(
      ondf$cpm %>% setNames(c("date", paste0("cpm_", names(.)[-1]))),
      ofdf$cpm %>% setNames(c("date", paste0("cpm_", names(.)[-1])))
    )
    impdf <- dplyr::inner_join(
      ondf$impression %>% setNames(c("date", paste0("imp_", names(.)[-1]))),
      ofdf$impression %>% setNames(c("date", paste0("imp_", names(.)[-1])))
    )

    # These come as pure tibbles
    prdf <- generatePriceData(fromDate, toDate, priceNames)
    didf <- generateDistributionData(fromDate, toDate, distributionNames)
    wedf <- generateWeatherData(fromDate, toDate, weatherNames)
    codf <- generateCompetitorData(fromDate, toDate, competitorNames)
    madf <- generateMacroData(fromDate, toDate, macroNames)
    # evdf <- generateEventData(fromDate, toDate, eventNames)

    mydf <- Reduce(
      function(x, y) dplyr::inner_join(x, y, by = "date"),
      list(mydf, wedf, codf, madf, didf, prdf, netdf, impdf, cpmdf)
    )
    mydf
  }
